c7n/actions/core.py

```diff
@@ -55,6 +57,71 @@ def factory(self, data, manager):
         return action_class(data, manager)
 
 
+class ActionResults:
+    AnnotationKey = "c7n:Actions"
+    allowed_states = ("ok", "skip", "error")
+
+    def __init__(self, action):
+        self.action = action
+        self.details = None
+        self.state = {}
+        self.metrics = {i: 0 for i in self.allowed_states}
+        self.resources = []
+
+    @property
+    def id_key(self):
+        return self.action.id_key
+
+    def initialize(self, resources):
+        self.resources = resources
+
+    def ok(self, resources):
+        self._add(resources, status="ok")
+
+    def skip(self, resources, reason):
+        self._add(resources, status="skip", reason=reason)
+
+    def error(self, resources, reason):
+        self._add(resources, status="error", reason=reason)
+
+    def exception(self, e):
```

Comment:
whats the delta between error and exception? .. exception classes don't per se clearly serialize